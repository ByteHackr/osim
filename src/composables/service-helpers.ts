import { getDisplayedOsidbError } from '@/services/OsidbAuthService';
import { useToastStore } from '@/stores/ToastStore';
import { AxiosError, type AxiosResponse } from 'axios';

export function createCatchHandler(title: string = 'Error', shouldThrow: boolean = true) {
  return (error: AxiosError) => {
    const displayedError = getDisplayedOsidbError(error);
    const { addToast } = useToastStore();
    addToast({
      title,
      body: `${displayedError}: ${error.response?.data}`,
      css: 'warning',
    });
    console.error('‚ùå ', error);
    if (shouldThrow) throw error;
    return error;
  };
}

export function createSuccessHandler({ title, body }: Record<string, string>, onSucceed: Function = () => {}) {
  return (response: AxiosResponse) => {
    const { addToast } = useToastStore();
    onSucceed(response);
    addToast({
      title,
      body,
      css: 'success',
    });
  };
}
