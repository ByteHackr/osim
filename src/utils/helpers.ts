import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';
import * as R from 'ramda';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );

type DeepMappable = any[] | Record<string, any>

const isNonEmptyArray = (value: any) => R.is(Array, value) && value.length > 0;
const isNonArrayObject = (value: any) => R.is(Object, value) && !R.is(Array, value);
const isDeepMappable = (value: DeepMappable) => isNonEmptyArray(value) || isNonArrayObject(value);

export const deepMap = (transform: (arg: any) => any, object: DeepMappable): any =>
  R.map(
    (val: any) => isDeepMappable(val)
      ? deepMap(transform, val)
      : transform(val)
    , object
  );

export const cveRegex = /^\s*(CVE-\d{4}-\d{4,7})\s*$/;
