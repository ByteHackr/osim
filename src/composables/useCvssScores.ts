import { putFlawCvssScores, postFlawCvssScores } from '@/services/FlawService';
import { type Flaw } from '@/generated-client';
import { createSuccessHandler, createCatchHandler } from './service-helpers';
import { computed, ref, watch, type Ref } from 'vue';

const onSucceed = createSuccessHandler({ title: 'Success!', body: 'Saved CVSS Scores' });
const onError = createCatchHandler('Error updating Flaw CVSS data');

export function useCvssScores(flaw: Ref<Flaw>) {
  const wasCvssModified = ref(false);
  const redHatCvssData = flaw.value.cvss_scores.find((assessment) => assessment.issuer === 'RH');
  const flawRhCvss = ref(redHatCvssData || flaw.value.cvss_scores[0]);
  
  watch(flawRhCvss, () => wasCvssModified.value = true, {deep: true});

  const nistCvssScore = computed(() => flaw.value.cvss_scores.find((score) => score.issuer === 'NIST'));
  const flawNvdCvssScore = computed(() => nistCvssScore.value?.score || flaw.value.nvd_cvss3);

  async function saveCvssScores() {
    if (flawRhCvss.value.created_dt) {
      return putFlawCvssScores(
        flaw.value.uuid,
        flawRhCvss.value.uuid || '',
        flawRhCvss.value as unknown
      )
        .then(onSucceed)
        .catch(onError);
    } else {
      const requestBody = {
        // "score":  is recalculated based on the vector by OSIDB and does not need to be included
        comment: flawRhCvss.value.comment,
        cvss_version: 'V3',
        issuer: 'RH',
        vector: flawRhCvss.value.vector,
        embargoed: flaw.value.embargoed,
      };
      return postFlawCvssScores(flaw.value.uuid, requestBody as unknown)
        .then(onSucceed)
        .catch(onError);
    }
  }

  return {
    wasCvssModified,
    flawNvdCvssScore,
    flawRhCvss,
    saveCvssScores,
  };
}

