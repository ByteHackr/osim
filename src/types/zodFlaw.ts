import { z } from 'zod';
import {
  FlawType,
  MajorIncidentStateEnum,
  NistCvssValidationEnum,
  RequiresSummaryEnum,
  Source642Enum,
  IssuerEnum,
  FlawMetaType,
  FlawReferenceType,
} from '../generated-client';
import { DateTime } from 'luxon';
import { cveRegex } from '@/utils/helpers';
import { zodOsimDateTime, ImpactEnumWithBlank, ZodFlawClassification } from './zodShared';
import { ZodAffectSchema, type ZodAffectType } from './zodAffect';

export const FlawTypeWithBlank = { '': '', ...FlawType } as const;
export const RequiresSummaryEnumWithBlank = { '': '', ...RequiresSummaryEnum } as const;
export const Source642EnumWithBlank = { '': '', ...Source642Enum } as const;
export const MajorIncidentStateEnumWithBlank = { '': '', ...MajorIncidentStateEnum } as const;
export const NistCvssValidationEnumWithBlank = { '': '', ...NistCvssValidationEnum } as const;

export const flawTypes = Object.values(FlawTypeWithBlank);
export const flawSources = Object.values(Source642EnumWithBlank);

const flawImpactsWeight = {
  '': 0,
  'LOW': 1,
  'MODERATE': 2,
  'IMPORTANT': 3,
  'CRITICAL': 4,
};

export const flawImpacts = Object.values(ImpactEnumWithBlank)
  .sort((a, b) => flawImpactsWeight[b] - flawImpactsWeight[a]);
export const flawIncidentStates = Object.values(MajorIncidentStateEnumWithBlank);
export const summaryRequiredStates = Object.values(RequiresSummaryEnumWithBlank);

export type ZodFlawAcknowledgmentType = z.infer<typeof FlawAcknowledgmentSchema>;
export const FlawAcknowledgmentSchema = z.object({
  name: z.string().default(''),
  affiliation: z.string().default(''),
  from_upstream: z.boolean().default(false),
  flaw: z.string().default(''),
  uuid: z.string().default(''),
  embargoed: z.boolean().default(false),
  created_dt: zodOsimDateTime().nullish(),
  updated_dt: zodOsimDateTime().nullish(),
});

export type ZodFlawReferenceType = z.infer<typeof FlawReferenceSchema>;
export const FlawReferenceSchema = z.object({
  description: z.string().default(''),
  uuid: z.string().default(''),
  type: z.nativeEnum(FlawReferenceType).default('ARTICLE'),
  url: z.string().default(''),
  embargoed: z.boolean().default(false),
  updated_dt: zodOsimDateTime().nullish().default(null),
}).superRefine((reference, zodContext) => {
  if (reference.type === 'ARTICLE' && !reference.url.match(/^https:\/\/access\.redhat\.com\//)) {
    zodContext.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'Red Hat Security Bulletin URLs must be from https://access.redhat.com/',
      path: ['url'],
    });
  }

  if (!reference.url.match(/^(?:https?:\/\/)/)) {
    zodContext.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'Reference URL must begin with either https:// or http://',
      path: ['url'],
    });
  }

  if (!reference.url) {
    zodContext.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'Reference URL cannot be empty',
      path: ['url'],
    });
  }

  if (!reference.description) {
    zodContext.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'Reference description cannot be empty',
      path: ['description'],
    });
  }
});

export const flawReferenceTypeValues = Object.values(FlawReferenceType);

export type ZodFlawCVSSType = z.infer<typeof FlawCVSSSchema>;
export type FlawCVSSSchemaType = typeof FlawCVSSSchema;
export const FlawCVSSSchema = z.object({
  comment: z.string().nullable(),
  cvss_version: z.string(),
  flaw: z.string().uuid(),
  issuer: z.nativeEnum(IssuerEnum),
  score: z.number(), // $float
  uuid: z.string().uuid().nullable(), // read-only
  vector: z.string().nullable(),
  embargoed: z.boolean().nullable(),
  created_dt: zodOsimDateTime().nullish(), // $date-time, // read-only
  updated_dt: zodOsimDateTime().nullish(), // $date-time,
});


export const ZodFlawCommentSchema = z.object({
  uuid: z.string(),
  type: z.string(),
  external_system_id: z.string(),
  order: z.number(),
  // meta_attr: z.record(z.string(), z.string().nullish()).nullish(),
  meta_attr: z.object({
    id: z.string().nullish(),
    tags: z.string().nullish(),
    text: z.string().nullish(),
    time: z.string().nullish(),
    count: z.string().nullish(),
    bug_id: z.string().nullish(),
    creator: z.string().nullish(),
    creator_id: z.string().nullish(),
    is_private: z
      .string()
      .transform((booleanString) => booleanString === 'True')
      .or(z.boolean())
      .nullish(),
    attachment_id: z.string().nullish(),
    creation_time: z.string().nullish(),
    private_groups: z
      .string()
      // .transform((jsonString: string) => JSON.parse(jsonString.replace(/'/g, '"')))
      .or(z.array(z.string()))
      .nullish(),
  }).nullish(),
  created_dt: zodOsimDateTime().nullish(),
  updated_dt: zodOsimDateTime().nullish(),
});

export type FlawMetaSchemaType = typeof ZodFlawMetaSchema;
export const ZodFlawMetaSchema = z.object({
  uuid: z.string().uuid(),
  type: z.nativeEnum(FlawMetaType).nullish(),
  meta_attr: z.record(z.string(), z.string().nullish()).nullish(),
  embargoed: z.boolean(), // read-only
  created_dt: zodOsimDateTime().nullish(), // $date-time,
  updated_dt: zodOsimDateTime().nullish(), // $date-time,
});

// const flawTypes: string[] = Object.values(FlawType);
export type ZodFlawType = z.infer<typeof ZodFlawSchema>;
export type FlawSchemaType = typeof ZodFlawSchema;

export const ZodFlawSchema = z.object({
  // type: z.nativeEnum(FlawType).optional(),
  type: z.nativeEnum(FlawTypeWithBlank).nullish(),
  uuid: z.string().default(''),
  cve_id: z.string().nullable().refine(
    (cve) => !cve || cveRegex.test(cve),
    {
      message: 'The CVE ID is invalid: It must begin with "CVE-", ' +
        'have a year between 1999-2999, and have an identifier at least ' +
        '4 digits long (e.g. use 0001 for 1). Please also check for ' +
        'unexpected characters like spaces.'
    }
  ),
  impact: z.nativeEnum(ImpactEnumWithBlank)
    .refine(
      Boolean,
      { message: 'You must select an impact before saving the Flaw.' }
    ),
  component: z.string().max(100).min(1),
  title: z.string().min(4),
  owner: z.string().nullish(),
  team_id: z.string().nullish(),
  trackers: z.array(z.string()).nullish(), // read-only
  classification: ZodFlawClassification.nullish(),
  description: z.string().refine(
    description => description.trim().length > 0,
    { message: 'Comment#0 cannot be empty.' }
  ),
  summary: z.string().nullish(),
  requires_summary: z.nativeEnum(RequiresSummaryEnumWithBlank).nullish(),
  statement: z.string().nullish(),
  cwe_id: z.string().max(255).nullish(),
  unembargo_dt: zodOsimDateTime().nullish(), // $date-time,
  reported_dt: zodOsimDateTime().nullish(), // $date-time,
  source: z.nativeEnum(Source642EnumWithBlank).refine(
    Boolean,
    { message: 'You must specify a source for this Flaw before saving.' }
  ),
  meta_attr: z.record(z.string(), z.string().nullish()).nullish(),
  mitigation: z.string().nullish(),
  cvss2: z.string().max(100).nullish(), // XXX deprecated
  cvss2_score: z.number().nullish(), // $float // XXX deprecated
  nvd_cvss2: z.string().max(100).nullish(), // XXX deprecated
  cvss3: z.string().max(100).nullish(), // XXX deprecated
  cvss3_score: z.number().nullish(), // $float // XXX deprecated
  nvd_cvss3: z.string().max(100).nullish(), // XXX deprecated
  major_incident_state: z.nativeEnum(MajorIncidentStateEnumWithBlank).nullish(),
  nist_cvss_validation: z.nativeEnum(NistCvssValidationEnumWithBlank).nullish(),
  affects: z.array(ZodAffectSchema), // read-only
  meta: z.array(ZodFlawMetaSchema),
  comments: z.array(ZodFlawCommentSchema),
  cvss_scores: z.array(FlawCVSSSchema),
  references: z.array(FlawReferenceSchema),
  acknowledgments: z.array(FlawAcknowledgmentSchema),
  embargoed: z.boolean(),
  updated_dt: zodOsimDateTime().nullish(), // $date-time,
}).superRefine((zodFlaw, zodContext) => {

  const raiseIssue = (message: string, path: string[]) => {
    zodContext.addIssue({
      code: z.ZodIssueCode.custom,
      message,
      path,
    });
  };

  const duplicatedAffects = (affects: ZodAffectType[]) => {
    const map: Record<string, boolean> = {};
    for (let i = 0; i < affects.length; i++) {
      const key = `${affects[i].ps_module}-${affects[i].ps_component}`;
      if (map[key]) {
        return {
          index: i,
        };
      } else {
        map[key] = true;
      }
    }
    return null;
  };

  if (
    zodFlaw.requires_summary
    && ['REQUESTED', 'APPROVED'].includes(zodFlaw.requires_summary)
    && zodFlaw.summary === ''
  ) {
    raiseIssue('Description cannot be blank if requested or approved.', ['summary']);
  }

  if (
    zodFlaw.requires_summary !== 'APPROVED'
    && zodFlaw.major_incident_state
    && ['APPROVED', 'CISA_APPROVED'].includes(zodFlaw.major_incident_state)
  ) {
    raiseIssue('Description must be approved for Major Incidents.', ['summary']);
  }

  const unembargo_dt = DateTime.fromISO(`${zodFlaw.unembargo_dt}`).toISODate();
  const reported_dt = DateTime.fromISO(`${zodFlaw.reported_dt}`).toISODate();

  if (zodFlaw.embargoed && unembargo_dt && unembargo_dt < DateTime.now().toISODate()) {
    raiseIssue('An embargoed flaw must have a public date in the future.', ['unembargo_dt']);
  }

  if (!zodFlaw.embargoed && !unembargo_dt) {
    // if embargoed and dt is not set, shows up the error
    // This behaviour is not acceptable because if a flaw is unembargoed,
    // it means the flaw is public already which requires the date when this was made public.
    raiseIssue('A public flaw must have a public date set.', ['unembargo_dt']);
  }

  if (!zodFlaw.embargoed && unembargo_dt && unembargo_dt > DateTime.now().toISODate()) {
    // if NOT embargoed and updated date is in the future,
    // it shows an error, to set the current date or an older date instead
    raiseIssue('A public flaw cannot have a public date in the future.', ['unembargo_dt']);
  }

  if (!zodFlaw.reported_dt) {
    raiseIssue('Reported date is required.', ['reported_dt']);
  }

  if (reported_dt && reported_dt > DateTime.now().toISODate()) {
    raiseIssue('Reported date cannot be in the future.', ['reported_dt']);
  }

  const affect = duplicatedAffects(zodFlaw.affects);
  if (affect) {
    raiseIssue(
      'Affected component cannot be registered on the affected module twice',
      [`Affects/${affect.index}/component`],
    );
  }
});
