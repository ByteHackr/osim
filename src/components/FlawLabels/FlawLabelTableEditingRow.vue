<script setup lang="ts">
import { StateEnum } from '@/generated-client';
import type { ZodFlawType } from '@/types';
import { FlawLabelTypeEnum } from '@/types/zodFlaw';
import { watchedRef } from '@/utils/helpers';

import FlawLabelsContributor from './FlawLabelsContributor.vue';

const { contextLabels, initalLabel } = defineProps<{
  contextLabels?: string[];
  initalLabel?: NonNullable<ZodFlawType['labels']>[number];
}>();

const emit = defineEmits<{
  cancel: [];
  save: [label: Omit<NonNullable<ZodFlawType['labels']>[number], 'uuid'>];
}>();

const [labelState, hasStateChanged] = watchedRef<StateEnum>(initalLabel?.state ?? StateEnum.New);
const [labelName, hasNameChanged] = watchedRef(initalLabel?.label ?? '');
const [labelContributor, hasContributorChanged] = watchedRef(initalLabel?.contributor ?? '');

const emitSave = () => {
  // if nothing changed, do not emit save
  if (
    !hasStateChanged.value
    && !hasNameChanged.value
    && !hasContributorChanged.value
  ) {
    return emit('cancel');
  }

  // if label name is empty, do nothing
  if (!labelName.value) {
    return;
  }

  emit('save', {
    ...initalLabel,
    state: labelState.value,
    label: labelName.value,
    contributor: labelContributor.value,
    relevant: initalLabel?.relevant ?? true,
    type: initalLabel?.type ?? FlawLabelTypeEnum.CONTEXT_BASED,
  });
};
</script>

<template>
  <td>
    <select v-model="labelState" class="form-select">
      <option
        v-for="state in StateEnum"
        :key="state"
        :value="state"
      >{{ state }}</option>
    </select>
  </td>
  <td
    :class="{
      'fw-bold': labelState === 'REQ',
      'text-decoration-line-through': !initalLabel?.relevant,
    }"
  >
    <template v-if="initalLabel?.type !== FlawLabelTypeEnum.PRODUCT_FAMILY">
      <select
        v-model="labelName"
        class="form-select"
        required
      >
        <option
          v-for="label in contextLabels"
          :key="label"
          :value="label"
        >{{ label }}</option>
      </select>
    </template>
    <template v-else>
      {{ initalLabel?.label }}
    </template>
  </td>
  <td>
    <FlawLabelsContributor v-model="labelContributor" />
  </td>
  <td>
    <div class="actions">
      <button
        type="button"
        class="btn btn-sm btn-dark"
        title="Save"
        @click="emitSave"
      >
        <i class="bi bi-check2" />
      </button>
      <button
        type="button"
        class="btn btn-sm btn-dark"
        title="Cancel"
        @click="emit('cancel')"
      >
        <i class="bi bi-x" />
      </button>
    </div>
  </td>
</template>

<style scoped lang="scss">
input,
select {
  padding: 0.15rem 0.5rem;
}
</style>
