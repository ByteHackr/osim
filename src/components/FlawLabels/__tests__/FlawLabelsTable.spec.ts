import { flushPromises } from '@vue/test-utils';

import { useFlawLabels } from '@/composables/useFlawLabels';

import { mountWithConfig } from '@/__tests__/helpers';
import { FlawLabelTypeEnum } from '@/types/zodFlaw';
import { StateEnum } from '@/generated-client';

import FlawLabelsTable from '../FlawLabelsTable.vue';
import FlawLabelTableEditingRow from '../FlawLabelTableEditingRow.vue';

vi.mock('@/services/LabelsService', () => ({
  fetchLabels: vi.fn().mockResolvedValue([
    { type: 'context_based', name: 'test' },
    { type: 'context_based', name: 'test1' },
  ]),
}));

const mountFlawLabelsTable = (props = {}) => mountWithConfig(FlawLabelsTable, {
  props: {
    modelValue: [
      { type: FlawLabelTypeEnum.CONTEXT_BASED, label: 'test', contributor: 'skynet', state: StateEnum.New },
    ],
    ...props,
  },
});

describe('flawLabelsTable', () => {
  it('should render flaw labels table', async () => {
    const wrapper = mountFlawLabelsTable();
    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle add label', async () => {
    const wrapper = mountFlawLabelsTable();
    const { areLabelsUpdated } = useFlawLabels();
    await flushPromises();

    await wrapper.find('.table-new-row td').trigger('click');
    const editRow = wrapper.findComponent(FlawLabelTableEditingRow);
    editRow.vm.$emit('save', {
      type: FlawLabelTypeEnum.CONTEXT_BASED,
      label: 'test1',
      contributor: 'skynet',
      state: StateEnum.New,
    });
    await flushPromises();

    expect(areLabelsUpdated.value).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle edit label', async () => {
    const wrapper = mountFlawLabelsTable();
    const { areLabelsUpdated } = useFlawLabels();
    await flushPromises();

    await wrapper.find('button[title="Edit label"]').trigger('click');
    const editRow = wrapper.findComponent(FlawLabelTableEditingRow);
    editRow.vm.$emit('save', {
      type: FlawLabelTypeEnum.CONTEXT_BASED,
      label: 'test1',
      contributor: 'skynet',
      state: StateEnum.New,
    });
    await flushPromises();

    expect(areLabelsUpdated.value).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle delete label', async () => {
    const wrapper = mountFlawLabelsTable();
    const { areLabelsUpdated } = useFlawLabels();
    await flushPromises();

    await wrapper.find('button[title="Delete label"]').trigger('click');
    await flushPromises();

    expect(areLabelsUpdated.value).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle undo delete label', async () => {
    const wrapper = mountFlawLabelsTable();
    const { areLabelsUpdated } = useFlawLabels();
    await flushPromises();

    await wrapper.find('button[title="Delete label"]').trigger('click');
    await flushPromises();

    await wrapper.find('button[title="Undo delete"]').trigger('click');
    await flushPromises();

    expect(areLabelsUpdated.value).toBe(false);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
