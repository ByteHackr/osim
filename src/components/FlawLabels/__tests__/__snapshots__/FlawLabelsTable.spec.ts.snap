// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flawLabelsTable > should handle add label 1`] = `
"<div data-v-4715e8e2="" data-v-5d530569="" class="osim-collapsible-label"><button data-v-4715e8e2="" type="button" class="me-2 osim-collapsible-toggle"><i data-v-4715e8e2="" class="bi me-1 bi-dash-square-dotted"></i><span data-v-5d530569="" class="section-label">Contributors</span></button>
  <div data-v-4715e8e2="" class="ps-3 border-start">
    <div data-v-4715e8e2="" class="">
      <table data-v-5d530569="" class="table table-striped table-hover">
        <thead data-v-5d530569="" class="table-dark">
          <tr data-v-5d530569="">
            <th data-v-5d530569="">State</th>
            <th data-v-5d530569="">Label</th>
            <th data-v-5d530569="">Contributor</th>
            <th data-v-5d530569="" class="table-actions-header">Actions</th>
          </tr>
        </thead>
        <tbody data-v-5d530569="">
          <tr data-v-5d530569="" class="">
            <td data-v-5d530569="" class="">NEW</td>
            <td data-v-5d530569="" class="text-decoration-line-through" title="Associated affect was removed">test</td>
            <td data-v-5d530569="">skynet</td>
            <td data-v-5d530569="">
              <div data-v-5d530569="" class="actions"><button data-v-5d530569="" type="button" title="Edit label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-pencil"></i></button>
                <!--v-if--><button data-v-5d530569="" type="button" title="Delete label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-trash"></i></button>
              </div>
            </td>
          </tr>
          <tr data-v-5d530569="" class="new">
            <td data-v-5d530569="" class="">NEW</td>
            <td data-v-5d530569="" class="text-decoration-line-through" title="Associated affect was removed">test1</td>
            <td data-v-5d530569="">skynet</td>
            <td data-v-5d530569="">
              <div data-v-5d530569="" class="actions"><button data-v-5d530569="" type="button" title="Edit label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-pencil"></i></button>
                <!--v-if--><button data-v-5d530569="" type="button" title="Delete label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-trash"></i></button>
              </div>
            </td>
          </tr>
          <!--v-if-->
          <!--v-if-->
        </tbody>
      </table>
    </div>
  </div>
</div>"
`;

exports[`flawLabelsTable > should handle delete label 1`] = `
"<div data-v-4715e8e2="" data-v-5d530569="" class="osim-collapsible-label"><button data-v-4715e8e2="" type="button" class="me-2 osim-collapsible-toggle"><i data-v-4715e8e2="" class="bi me-1 bi-dash-square-dotted"></i><span data-v-5d530569="" class="section-label">Contributors</span></button>
  <div data-v-4715e8e2="" class="ps-3 border-start">
    <div data-v-4715e8e2="" class="">
      <table data-v-5d530569="" class="table table-striped table-hover">
        <thead data-v-5d530569="" class="table-dark">
          <tr data-v-5d530569="">
            <th data-v-5d530569="">State</th>
            <th data-v-5d530569="">Label</th>
            <th data-v-5d530569="">Contributor</th>
            <th data-v-5d530569="" class="table-actions-header">Actions</th>
          </tr>
        </thead>
        <tbody data-v-5d530569="">
          <tr data-v-5d530569="" class="deleted">
            <td data-v-5d530569="" class="">NEW</td>
            <td data-v-5d530569="" class="text-decoration-line-through" title="Associated affect was removed">test</td>
            <td data-v-5d530569="">skynet</td>
            <td data-v-5d530569="">
              <div data-v-5d530569="" class="actions">
                <!--v-if--><button data-v-5d530569="" type="button" title="Undo delete" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-arrow-counterclockwise"></i></button>
                <!--v-if-->
              </div>
            </td>
          </tr>
          <!--v-if-->
          <tr data-v-5d530569="" class="table-new-row">
            <td data-v-5d530569="" colspan="100%" class="text-center" title="Add new label"><i data-v-5d530569="" class="bi bi-plus"></i></td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>"
`;

exports[`flawLabelsTable > should handle edit label 1`] = `
"<div data-v-4715e8e2="" data-v-5d530569="" class="osim-collapsible-label"><button data-v-4715e8e2="" type="button" class="me-2 osim-collapsible-toggle"><i data-v-4715e8e2="" class="bi me-1 bi-dash-square-dotted"></i><span data-v-5d530569="" class="section-label">Contributors</span></button>
  <div data-v-4715e8e2="" class="ps-3 border-start">
    <div data-v-4715e8e2="" class="">
      <table data-v-5d530569="" class="table table-striped table-hover">
        <thead data-v-5d530569="" class="table-dark">
          <tr data-v-5d530569="">
            <th data-v-5d530569="">State</th>
            <th data-v-5d530569="">Label</th>
            <th data-v-5d530569="">Contributor</th>
            <th data-v-5d530569="" class="table-actions-header">Actions</th>
          </tr>
        </thead>
        <tbody data-v-5d530569="">
          <tr data-v-5d530569="" class="">
            <td data-v-5d530569="" class="">NEW</td>
            <td data-v-5d530569="" class="text-decoration-line-through" title="Associated affect was removed">test</td>
            <td data-v-5d530569="">skynet</td>
            <td data-v-5d530569="">
              <div data-v-5d530569="" class="actions"><button data-v-5d530569="" type="button" title="Edit label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-pencil"></i></button>
                <!--v-if--><button data-v-5d530569="" type="button" title="Delete label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-trash"></i></button>
              </div>
            </td>
          </tr>
          <tr data-v-5d530569="" class="updated">
            <td data-v-5d530569="" class="">NEW</td>
            <td data-v-5d530569="" class="text-decoration-line-through" title="Associated affect was removed">test1</td>
            <td data-v-5d530569="">skynet</td>
            <td data-v-5d530569="">
              <div data-v-5d530569="" class="actions"><button data-v-5d530569="" type="button" title="Edit label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-pencil"></i></button>
                <!--v-if--><button data-v-5d530569="" type="button" title="Delete label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-trash"></i></button>
              </div>
            </td>
          </tr>
          <!--v-if-->
          <!--v-if-->
        </tbody>
      </table>
    </div>
  </div>
</div>"
`;

exports[`flawLabelsTable > should handle undo delete label 1`] = `
"<div data-v-4715e8e2="" data-v-5d530569="" class="osim-collapsible-label"><button data-v-4715e8e2="" type="button" class="me-2 osim-collapsible-toggle"><i data-v-4715e8e2="" class="bi me-1 bi-dash-square-dotted"></i><span data-v-5d530569="" class="section-label">Contributors</span></button>
  <div data-v-4715e8e2="" class="ps-3 border-start">
    <div data-v-4715e8e2="" class="">
      <table data-v-5d530569="" class="table table-striped table-hover">
        <thead data-v-5d530569="" class="table-dark">
          <tr data-v-5d530569="">
            <th data-v-5d530569="">State</th>
            <th data-v-5d530569="">Label</th>
            <th data-v-5d530569="">Contributor</th>
            <th data-v-5d530569="" class="table-actions-header">Actions</th>
          </tr>
        </thead>
        <tbody data-v-5d530569="">
          <tr data-v-5d530569="" class="">
            <td data-v-5d530569="" class="">NEW</td>
            <td data-v-5d530569="" class="text-decoration-line-through" title="Associated affect was removed">test</td>
            <td data-v-5d530569="">skynet</td>
            <td data-v-5d530569="">
              <div data-v-5d530569="" class="actions"><button data-v-5d530569="" type="button" title="Edit label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-pencil"></i></button>
                <!--v-if--><button data-v-5d530569="" type="button" title="Delete label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-trash"></i></button>
              </div>
            </td>
          </tr>
          <!--v-if-->
          <tr data-v-5d530569="" class="table-new-row">
            <td data-v-5d530569="" colspan="100%" class="text-center" title="Add new label"><i data-v-5d530569="" class="bi bi-plus"></i></td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>"
`;

exports[`flawLabelsTable > should render flaw labels table 1`] = `
"<div data-v-4715e8e2="" data-v-5d530569="" class="osim-collapsible-label"><button data-v-4715e8e2="" type="button" class="me-2 osim-collapsible-toggle"><i data-v-4715e8e2="" class="bi me-1 bi-dash-square-dotted"></i><span data-v-5d530569="" class="section-label">Contributors</span></button>
  <div data-v-4715e8e2="" class="ps-3 border-start">
    <div data-v-4715e8e2="" class="">
      <table data-v-5d530569="" class="table table-striped table-hover">
        <thead data-v-5d530569="" class="table-dark">
          <tr data-v-5d530569="">
            <th data-v-5d530569="">State</th>
            <th data-v-5d530569="">Label</th>
            <th data-v-5d530569="">Contributor</th>
            <th data-v-5d530569="" class="table-actions-header">Actions</th>
          </tr>
        </thead>
        <tbody data-v-5d530569="">
          <tr data-v-5d530569="" class="">
            <td data-v-5d530569="" class="">NEW</td>
            <td data-v-5d530569="" class="text-decoration-line-through" title="Associated affect was removed">test</td>
            <td data-v-5d530569="">skynet</td>
            <td data-v-5d530569="">
              <div data-v-5d530569="" class="actions"><button data-v-5d530569="" type="button" title="Edit label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-pencil"></i></button>
                <!--v-if--><button data-v-5d530569="" type="button" title="Delete label" class="btn btn-sm btn-dark"><i data-v-5d530569="" class="bi bi-trash"></i></button>
              </div>
            </td>
          </tr>
          <!--v-if-->
          <tr data-v-5d530569="" class="table-new-row">
            <td data-v-5d530569="" colspan="100%" class="text-center" title="Add new label"><i data-v-5d530569="" class="bi bi-plus"></i></td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>"
`;
